WIP: authentication: direct external OIDC provider

liouk opened the pull request:
This PR adds an enhancement for the implementation of using a direct external OIDC provider instead of the built-in oauth stack in OCP.

See also https://issues.redhat.com/browse/OCPSTRAT-306.

liouk commented:
/retest

liouk commented:
/retest

stlaz commented on enhancements/authentication/direct-external-oidc-provider.md:
````diff
@@ -0,0 +1,409 @@
+}
+```
+
+For more details on the `OIDCProvider` type and its fields, see [here](https://github.com/openshift/api/blob/master/config/v1/types_authentication.go#L197).
````
I wouldn't point to `master`

liouk replied:
:+1: changed it to point to the commit that introduces the new fields: https://github.com/openshift/api/blob/fa2f9ad8645efed0a83c24de025fd7fe791cc558/config/v1/types_authentication.go

stlaz commented on enhancements/authentication/direct-external-oidc-provider.md:
```diff
@@ -0,0 +1,409 @@
+
+#### Single-node Deployments or MicroShift
+
+Configuring an external OIDC provider in a Single-Node deployment of OpenShift (SNO) will result in reduced resource consumption overall, due to the fact that once an external provider is configured successfully, the system will deactivate/remove components and resources that are unused (e.g. the oauth-server and oauth-apiserver pods won't run anymore).
```
AFAIK MicroShift doesn't run with multiple users and it does not even deploy the integrated oauth-server

liouk commented on enhancements/authentication/direct-external-oidc-provider.md:
```diff
@@ -0,0 +1,409 @@
+
+## Proposal
+
+This proposal introduces changes to components that act as clients to the existing built-in OAuth server, and to the cluster-authentication-operator which manages the OAuth stack. The built-in OAuth server will still be available as the default option; the user will be able to configure their provider as a Day-2 configuration.
```
I'll rephrase -- "and to the CAO" was supposed to mean "and changes to the CAO", i.e. not the clients part. :+1:

liouk commented on enhancements/authentication/direct-external-oidc-provider.md:
```diff
@@ -0,0 +1,409 @@
+
+The configuration details required to connect to an external OIDC provider are:
+- the provider URL
+- the ID and secret of the corresponding OIDC client at the provider's side
```
I'll loosen the wording to take care of this and the previous comment.

stlaz replied:
still to be updated

stlaz commented on enhancements/authentication/direct-external-oidc-provider.md:
```diff
@@ -0,0 +1,222 @@
+
+### Non-Goals
+
+n/a
```
\+ keep the integrated OpenShift authentication stack working

stlaz commented on enhancements/authentication/direct-external-oidc-provider.md:
```diff
@@ -0,0 +1,222 @@
+Currently, any component that needs to obtain tokens or authenticate users does so against the built-in OAuth server. In order to integrate and use an external OIDC Identity Provider directly, any client component must replace its configuration to call the external OIDC provider instead of the built-in server. The components that act as OAuth server clients are:
+- OpenShift Console calls the OAuth server for user login to the console, and to obtain and display API access tokens
+- `oc` calls the OAuth server for user login to the API, and to obtain API access tokens
+- kube-apiserver calls the OAuth server via an authentication webhook for token management
```
the enhancement statement is not true - the KAS reaches out to the oauth-apiserver

stlaz commented on enhancements/authentication/direct-external-oidc-provider.md:
```diff
@@ -0,0 +1,409 @@
+Currently, any component that needs to obtain tokens or authenticate users does so against the built-in OAuth server. In order to integrate and use an external OIDC Identity Provider directly, any client component must replace its configuration to call the external OIDC provider instead of the built-in server. The components that act as OAuth server clients are:
+- OpenShift Console calls the OAuth server for user login to the console, and to obtain and display API access tokens
+- `oc` calls the OAuth server for user login to the API, and to obtain API access tokens
+- kube-apiserver calls the OAuth server via an authentication webhook for token management
```
you still did not incorporate the above comment

stlaz commented on enhancements/authentication/direct-external-oidc-provider.md:
```diff
@@ -0,0 +1,222 @@
+- the `AuthMetadata` observer sets the `OauthMetadataFile` field of the CR with the path for a ConfigMap referenced by the authentication config
+
+The operator must watch the Authentication CR for changes, and when it detects an external OIDC provider configuration, it must make the following changes, in order to update its configuration to use the external provider:
+- the `WebhookTokenAuthenticator` observer must replace the secret used as a webhook token authenticator for the API server with the one of the new provider, and synchronize it to the `openshift-kube-apiserver` namespace
```
This doesn't make sense. What would you even use the secret for?

stlaz commented on enhancements/authentication/direct-external-oidc-provider.md:
```diff
@@ -0,0 +1,222 @@
+
+When the built-in OAuth server is used for authentication (the default and original cluster state), the cluster-authentication-operator manages all controllers and resources related to it; notably, it manages the deployments of the oauth-server and oauth-apiserver, and manages resources such as the respective namespaces, service accounts, role bindings, services, OAuth clients, etc. Moreover, it  ...
+
+In case an external OIDC provider is configured for authentication, then these controllers and resources are not useful or irrelevant. Therefore, the operator must watch the Authentication CR, and when it detects a valid external OIDC provider configuration, it must stop its controllers and remove its resources. The operator will not be monitoring the oauth-server and oauth-apiserver any longer,  ...
```
you probably don't want the controllers to be actually stopped but rather turn them into a state where they remove the state they otherwise push to the cluster.

stlaz commented on enhancements/authentication/direct-external-oidc-provider.md:
```diff
@@ -0,0 +1,222 @@
+
+### Risks and Mitigations
+
+Configuring an external OIDC identity provider for authentication by definition means that any security updates or patches must be managed independently from the cluster itself, i.e. cluster updates will not resolve security issues relevant to the provider itself; the provider will have to be updated separately. Additionally, new functionality or features on the provider's side might need integra ...
```
what about losing all your OAuth and User API data?

liouk commented:
/remove-lifecycle rotten

deads2k commented on enhancements/authentication/direct-external-oidc-provider.md:
```diff
@@ -0,0 +1,225 @@
+
+## Graduation Criteria
+
+The goal for this enhancement is to go directly to GA, because the feature is already available and in GA in HCP.
```
it will continue through techpreview, but I'm hopeful we do it quickly.

deads2k commented on enhancements/authentication/direct-external-oidc-provider.md:
```diff
@@ -0,0 +1,225 @@
+Overall, for this feature there must be e2e tests that cover the following:
+- configuring an external OIDC provider on a cluster that uses the built-in OAuth stack (good/bad configurations should be tested)
+- on a cluster that uses an external OIDC provider, test reverting configuration back to the built-in OAuth stack (good/bad configurations should be tested)
+- on a cluster that uses an external OIDC provider, test monitoring and cluster-authentication-operator status when the provider becomes unavailable
```
need to be sure that the user mapping capabilities work as well.

deads2k commented on enhancements/authentication/direct-external-oidc-provider.md:
```diff
@@ -0,0 +1,225 @@
+
+### Implementation Details/Notes/Constraints
+
+#### cluster-kube-apiserver-operator
```
should these trigger kube-apiserver revisions or simply trigger an all-at-once instant rollout using live file reloads for the kube-apiserver?

liouk replied:
We've agreed that this decision must be documented in this enhancement. Putting a hold until this is resolved.

/hold

deads2k commented:
looks like a good overall direction. I'll approve, but leave the lgtm with the team

/approve

liouk commented:
Added a fixup (818c4bba7b28be846c8d0415744b4c290fa64471) that addresses the comments above, for easier further review.

Adding a hold until fixups are squashed.
/hold

liouk commented on enhancements/authentication/direct-external-oidc-provider.md:
```diff
@@ -0,0 +1,238 @@
+	// for this cluster
+	// Can only be set if "Type" is set to "OIDC".
+	//
+	// At most one provider can be configured.
```
Future-proofing when designing the API.

liouk commented on enhancements/authentication/direct-external-oidc-provider.md:
````diff
@@ -0,0 +1,238 @@
+}
+```
+
+For more details on the `OIDCProvider` type and its fields, see [here](https://github.com/openshift/api/blob/fa2f9ad8645efed0a83c24de025fd7fe791cc558/config/v1/types_authentication.go#L197).
````
This refers to the *type* name, which is `OIDCProvider` indeed.

liouk commented on enhancements/authentication/direct-external-oidc-provider.md:
```diff
@@ -0,0 +1,238 @@
+
+In case an external OIDC provider is configured for authentication, then these controllers and resources are neither useful nor relevant. Therefore, the operator must watch the Authentication CR, and when it detects a valid external OIDC provider configuration, it must turn its controllers into a state where they remove the state/resources they otherwise push to the cluster. The operator will not ...
+
+Note that the operator must first validate the new provider configuration before proceeding with the deactivation of the built-in OAuth stack. Also, in case the authentication configuration gets changed back to the built-in OAuth server, the operator must revert these changes and bring its operands and relevant resources back to the original state of affairs.
```
I'm thinking we can use the Status of the Authentication CR so that the KAS-o can post its OIDC config status. The trigger for teh CAO to start cleanup will be two conditions: auth type is OIDC, and the KAS-o has successfully configured OIDC.

liouk commented on enhancements/authentication/direct-external-oidc-provider.md:
```diff
@@ -0,0 +1,238 @@
+
+Enabling an external OIDC provider to an OCP cluster will result in the oauth-apiserver being removed from the system; this inherently means that the two API Services it is serving (`v1.oauth.openshift.io`, `v1.user.openshift.io`) will be gone from the cluster, and therefore any related data will be lost. It is the user's responsibility to create backups of any required data.
+
+Additionally, configuring an external OIDC identity provider for authentication by definition means that any security updates or patches must be managed independently from the cluster itself, i.e. cluster updates will not resolve security issues relevant to the provider itself; the provider will have to be updated separately. Additionally, new functionality or features on the provider's side migh ...
```
added content about this scenario

liouk commented:
Changing to WIP in order to clarify the open questions.

/retitle [WIP] authentication: direct external OIDC provider

p0lyn0mial commented on enhancements/authentication/direct-external-oidc-provider.md:
```diff
@@ -0,0 +1,316 @@
+2. this means that the `spec.oauthMetadata` field will be set to empty; however `status.integratedOAuthMetadata` will still have a value
+3. KAS-o config observer picks up the auth type change, configures KAS pods for OIDC, triggers a rollout
+4. once rollout is successful, CAO picks up the change to OIDC, removes oauth-specific resources, including `status.integratedOAuthMetadata`
+5. KAS-o oauth-metadata config observer picks up the change in `status.integratedOAuthMetadata` and since `spec.oauthMetadata` is also unset, removes the oauth-metadata configmap; this results in a new config, which triggers a new rollout
```
mhm, I think that if you configure the external OIDC and remove the destination CM in the same step, no additional rollout will be required after cleaning up the status field, right?

p0lyn0mial replied:
Actually, thinking more about it, I believe we cannot guarantee that these two operations will be performed atomically.

p0lyn0mial replied:
What happens after configuring the external OIDC server? Does kas use the new one, or both?

liouk replied:
> if you configure the external OIDC and remove the destination CM in the same step, no additional rollout will be required after cleaning up the status field, right?
Right -- that's what I've tried to achieve with the alternative proposal. 

> we cannot guarantee that these two operations will be performed atomically
If I understand this correctly, we could achieve atomicity if we use the same config observer to configure either the webhook or OIDC depending on the auth type, in a mutually exclusive way, i.e. configuring the one will remove configuration of the other. Indeed we cannot guarantee this will happen atomically if we use different observers.

> Does kas use the new one, or both?
If we keep the webhook in the configuration and configure OIDC as well, both methods will be part of the authentication chain. If the first one in the chain fails to validate a token, then the next one will be used until it gets validated or the chain is finished. So both can be active at the same time.

One benefit of doing that, would be that we can keep the webhook up until the KAS-o has validated that OIDC works as expected; then it can mark its auth status as available, and the CAO can then clean up the oauth metadata and webhook (thus triggering a second rollout). If there's any problem with OIDC, the KAS-o won't mark OIDC as available, the CAO won't take down the webhook, and thus the webhook will keep working until the user fixes the OIDC config.

liouk replied:
Coming back to this for some further details I learned.

> If we keep the webhook in the configuration and configure OIDC as well, both methods will be part of the authentication chain.

While in theory this could be done, in practice that's not really possible due to our authentication CRD validation. The webhook can be defined only if type is _not_ OIDC, which means that if we switch configuration to OIDC in the auth CR then we have to get rid of the webhook.

liouk commented:
Fixups have been squashed, and open questions addressed; cancelling hold.

/hold cancel

xingxingxia commented on enhancements/authentication/direct-external-oidc-provider.md:
```diff
@@ -0,0 +1,293 @@
+
+#### Authentication Resource
+
+The cluster's Authentication CR (`authentication.config/cluster`) must be modified and updated with the configuration of the external OIDC provider in the `OIDCProviders` field.
```
```suggestion
The cluster's Authentication CR (`authentication.config/cluster`) must be modified and updated with the configuration of the external OIDC provider in the `OIDCProviders` field. The `WebhookTokenAuthenticator` field should be removed (by setting it `null`).
```
I add this explicitly because readers who are not aware of this may just set `oidcProviders` and ignore `webhookTokenAuthenticator`. This makes readers who attempt to run `oc patch` or `oc edit` on `authentication.config/cluster` the first time get:
```
The Authentication "cluster" is invalid: spec.webhookTokenAuthenticator: Invalid value: v1.WebhookTokenAuthenticator{KubeConfig:v1.SecretNameReference{Name:"webhook-authentication-integrated-oauth"}}: this field cannot be set with the "OIDC" .spec.type
```

ibihim reviewed (commented):
/lgtm
